src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceCodeRefactoringProvider.State.cs

```diff
@@ -0,0 +1,296 @@
+            /// <summary>
+            /// This is the new name we want to change the namespace to use.
+            /// Empty string means global namespace, whereas null means renaming action is not availalbe.
+            /// </summary>
+            public string TargetNamespace { get; }
+
+            public string QualifiedIdentifierFromDeclaration { get; }
+
+            public string QualifiedIdentifierFromFolders { get; }
+
+            private State(
+                Document document,
+                string rootNamespce,
+                string qualifiedIdentifierFromFolders,
+                string targetNamespace,
+                string declaredNamespace,                           
+                string qualifiedIdentifierFromDeclaration)
+            {                      
+                Document = document;
+                RootNamespace = rootNamespce;
+                QualifiedIdentifierFromFolders = qualifiedIdentifierFromFolders;
+                TargetNamespace = targetNamespace;
+                DeclaredNamespace = declaredNamespace;
+                QualifiedIdentifierFromDeclaration = qualifiedIdentifierFromDeclaration;
+            }
+
+            public static async Task<State> CreateAsync(TService service, Document document, TextSpan textSpan, CancellationToken cancellationToken)
+            {
+                if (!textSpan.IsEmpty)
+                {
+                    return null;
+                }
+
+                var workspace = document.Project.Solution.Workspace;
+                if (workspace.Kind == WorkspaceKind.MiscellaneousFiles)
+                {
+                    return null;
+                }
+
+                if (document.IsGeneratedCode(cancellationToken))
+                {
+                    return null;
+                }
+
+                // Ignore linked documents
+                // TODO: this only detect documents linked multiple times, need to figure out how to detect a document is linked, e.g. using <Link>, or shared project, etc.
+                if (document.GetLinkedDocumentIds().Any())
+                {
+                    return null;
+                }
+
+                var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+                // Bail if multiple namespaces are declared in the document (including nested declarations)
+                if (root.DescendantNodes().OfType<TNamespaceDeclarationSyntax>().Count() > 1)
+                {
+                    return null;
+                }
+
+                // Bail if cursor is not on namespace declaration.
+                // TODO: Note that this also means we will ignore document with only global namespace.
+                var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+                if (!syntaxFacts.IsOnNamespaceDeclaration(root, textSpan.Start))
```

Comment:
this is not an appropriate SyntaxFact.  It sounds like a feature-specific check, in which case the code for it should be in this feature.